enhancements/cloud-integration/infrastructure-external-platform-type.md

```diff
@@ -0,0 +1,513 @@
+
+### Workflow Description
+
+During [phase 1](#implementation-phases) proposed changes intended to have no effect and the "External"
+platform type should be handled the same as the "None", so no specific user interaction is expected.
+
+### API Extensions
+
+A new optional constant of `PlatformType` type will be added to the "openshift/api".
+
+```go
+const (
+    ...
+    // ExternalPlatformType represent generic infrastructure provider. Provider-specific components should be supplemented separately.
+    ExternalPlatformType PlatformType = "External"
+    ...
+)
+```
+
+Additionally, the respective external platform spec and status should be added to the infrastructure resource.
+
+```go
+type CloudControllerManagerMode string
+
+const (
+    // Cloud Controller Manager is enabled and expected to be supplied.
+    // Signaling MCO to set `--cloud-provider=external` flag to the kubelets.
+    CloudControllerManagerEnabled CloudControllerManagerMode = "Enabled"
+    // No Cloud Controller Manager is expected to be supplied.
+    // Signaling MCO not to set `--cloud-provider` flag to the kubelets.
+    CloudControllerManagerDisabled CloudControllerManagerMode = "Disabled"
+)
+
+type CloudControllerManagerSettings struct {
+    // Mode determines whether or not an external Cloud Controller Manager is expected to 
+    // be presented in the cluster.
+    // For engaging an external Cloud Controller Manager, certain flags are expected to be set to the kubelets.
+    // https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+    //
+    // When enabled, the respective operator (machine config operator) should set `--cloud-provider=external` flag to the kubelet.
+    // When omitted or disabled, no `cloud-provider` flag should be set.
+    // +kubebuilder:validation:Enum=Enabled;Disabled
+    // +optional
+    Mode CloudControllerManagerMode `json:"mode,omitempty"`
+}
+
+
+// ExternalPlatformSpec holds the desired state for the generic External infrastructure provider.
+type ExternalPlatformSpec struct{
+    // CloudControllerManager contains settings specific to the external Cloud Controller Manager (a.k.a. CCM or CPI)
+    // +optional
+    CloudControllerManager CloudControllerManagerSettings `json:"cloudControllerManager,omitempty"`
+}
+
+type PlatformSpec struct {
+    ...
+    // External contains settings specific to the generic External infrastructure provider.
+    // +optional
+    External *ExternalPlatformSpec `json:"external,omitempty"`
+}
+```
+
+```go
+
+// ExternalPlatformStatus holds the current status of the generic External infrastructure provider.
+type ExternalPlatformStatus struct{
+    // ProviderName holds the arbitrary string represented cloud provider name, expected to be set at the installation time.
+    // Intended to serve only for informational purposes and not expected to be used for decision-making.
+    // +kubebuilder:default:="Unknown"
```

Comment:
If this is in status how will it be set?

Reply:
good point, i think this needs to be in spec

Reply:
We can make it immutable if we want, for example, it could have some default value (eg `Unkown`) and we allow only a 1 way transition, from Unknown to another string, once set to a value that is not unknown, it becomes immutable

The alternative is to keep it in status and work out how the installer would set this value, it already populates the status of the Infra object today, it could be it just does that from the manifests dir in which case we may be able to get some of this for free

Reply:
talked with @lobziik a little today about this, the idea behind putting this in the status is that it will be set by the installer and then the user should not change it for the life of the cluster.

Reply:
In that case, are the installer folks aware that they will need to be setting this? How will a provider inject the correct string during the installation process. I'd like to see the mechanics behind this documented

Note, just because something is in status doesn't mean a user can't manipulate it